File: gateway/src/main/java/com/beat/gateway/guest/internal/GuestPasswordHashService.java
package com.beat.gateway.guest.internal;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;

import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;

import com.beat.contracts.auth.guest.GuestPasswordHashPort;

public final class GuestPasswordHashService implements GuestPasswordHashPort {

	private static final String BCRYPT_PREFIX = "$2";

	private final PasswordEncoder passwordEncoder = new BCryptPasswordEncoder();

	@Override
	public String encode(String rawPassword) {
		return passwordEncoder.encode(rawPassword);
	}

	@Override
	public boolean matches(String rawPassword, String storedPassword) {
		if (storedPassword == null || storedPassword.isBlank()) {
			return false;
		}
		if (isBcrypt(storedPassword)) {
			return passwordEncoder.matches(rawPassword, storedPassword);
		}
		return MessageDigest.isEqual(
			rawPassword.getBytes(StandardCharsets.UTF_8),
			storedPassword.getBytes(StandardCharsets.UTF_8)
		);
	}

	@Override
	public boolean needsUpgrade(String storedPassword) {
		return !isBcrypt(storedPassword) || passwordEncoder.upgradeEncoding(storedPassword);
	}

	private boolean isBcrypt(String storedPassword) {
		return storedPassword.startsWith(BCRYPT_PREFIX);
	}
}

File: gateway/src/main/kotlin/com/beat/gateway/guest/internal/GuestPasswordHashService.kt
package com.beat.gateway.guest.internal

import com.beat.contracts.auth.guest.GuestPasswordHashPort
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import java.nio.charset.StandardCharsets
import java.security.MessageDigest

class GuestPasswordHashService : GuestPasswordHashPort {

    private val passwordEncoder: PasswordEncoder = BCryptPasswordEncoder()

    override fun encode(rawPassword: String): String = passwordEncoder.encode(rawPassword)!!

    override fun matches(rawPassword: String, storedPassword: String): Boolean {
        if (storedPassword.isBlank()) {
            return false
        }
        if (isBcrypt(storedPassword)) {
            return passwordEncoder.matches(rawPassword, storedPassword)
        }
        return MessageDigest.isEqual(
            rawPassword.toByteArray(StandardCharsets.UTF_8),
            storedPassword.toByteArray(StandardCharsets.UTF_8),
        )
    }

    override fun needsUpgrade(storedPassword: String): Boolean =
        !isBcrypt(storedPassword) || passwordEncoder.upgradeEncoding(storedPassword)

    private fun isBcrypt(storedPassword: String): Boolean = storedPassword.startsWith(BCRYPT_PREFIX)

    companion object {
        private const val BCRYPT_PREFIX = "\$2"
    }
}
